### PR TITLE
[CICD] Update sonar job

### PR DESCRIPTION
**Pull request by ameijer**
## What does this PR change?
* uses a new approach for running sonarcloud. Fixes issues with baseline establishment and quality gating

## Does this PR relate to any other PRs?
* No

## How will this PR impact users?
* Improves quality by asking for unit testing of contributions

## Does this PR address any GitHub or Zendesk issues?
* Closes ...

## How was this PR tested?
* Tested in KCM

## Does this PR require changes to documentation?
* Maybe the contribution guide

## Have you labeled this PR and its corresponding Issue as "next release" if it should be part of the next OpenCost release? If not, why not?
* not a core feature

### CODE REVIEW

File: .github/workflows/sonar.yaml [hunk 1]
```diff
@@ -8,43 +8,71 @@ jobs:
...
-               archive_format: 'zip',
-            });
-            let fs = require('fs');
-            fs.writeFileSync(`${process.env.GITHUB_WORKSPACE}/oc-code-coverage.zip`, Buffer.from(download.data));
-      - name: 'Unzip code coverage'
-        run: unzip oc-code-coverage.zip -d coverage
-      - name: set env vars 
+          name: code-coverage
+          run-id: ${{ github.event.workflow_run.id }}
+          github-token: ${{ github.token }}
```
Comment (avivkeller) [inline .github/workflows/sonar.yaml:33]:
```suggestion
          github-token: ${{ github.token }}
          path: pr-artifact
```

File: .github/workflows/sonar.yaml [hunk 2]
```diff
@@ -8,43 +8,71 @@ jobs:
...
+      - name: Validate Coverage Vars
+        id: validate-vars
+        if: github.event.workflow_run.head_branch != 'develop'
+        shell: bash
+        run: | 
+          # check the PR number
+          pr_content=$(cat pr_num.txt | tr -d '\n' | tr -d ' ')
```
Comment (avivkeller) [inline .github/workflows/sonar.yaml]:
```suggestion
          pr_content=$(cat pr-artifact/pr_num.txt | tr -d '\n' | tr -d ' ')
```

File: .github/workflows/sonar.yaml [hunk 3]
```diff
@@ -8,43 +8,71 @@ jobs:
...
+
+          # Check if the content matches a single number
+          if [[ "$pr_content" =~ ^[0-9]+$ ]]; then
+            echo "The file 'pr_num.txt' contains a single number: $pr_content"
+          else
+            echo "The file 'pr_num.txt' does not contain a single number."
+            exit 1
+          fi
+
+          base_content=$(cat base.txt | tr -d '\n' | tr -d ' ')
```
Comment (avivkeller) [inline .github/workflows/sonar.yaml]:
```suggestion
          base_content=$(cat pr-artifact/base.txt | tr -d '\n' | tr -d ' ')
```

File: .github/workflows/sonar.yaml [hunk 4]
```diff
@@ -8,43 +8,71 @@ jobs:
...
+          if git check-ref-format --allow-onelevel "$base_content"; then
+            echo "The file 'base.txt' contains a valid git ref: $base_content"
+          else
+            echo "The file 'base.txt' does not contain a valid git ref: $base_content"
+            exit 1
+          fi
+
+          head_content=$(cat head.txt | tr -d '\n' | tr -d ' ')
```
Comment (avivkeller) [inline .github/workflows/sonar.yaml]:
```suggestion
          head_content=$(cat pr-artifact/head.txt | tr -d '\n' | tr -d ' ')
```

File: .github/workflows/sonar.yaml [hunk 5]
```diff
@@ -8,43 +8,71 @@ jobs:
...
+
+      - name: set vars 
+        id: set-vars
         run: | 
-          echo "SONAR_PR_NUM=$(cat coverage/pr_num.txt)" >> $GITHUB_ENV
-          echo "SONAR_BASE=$(cat coverage/base.txt)" >> $GITHUB_ENV
-          echo "SONAR_HEAD=$(cat coverage/head.txt)" >> $GITHUB_ENV
+          echo "SONAR_PR_NUM=$(cat pr_num.txt | tr -d '\n' | tr -d ' ')" >> $GITHUB_OUTPUT
+          echo "SONAR_BASE=$(cat base.txt | tr -d '\n' | tr -d ' ')" >> $GITHUB_OUTPUT
+          echo "SONAR_HEAD=$(cat head.txt | tr -d '\n' | tr -d ' ')" >> $GITHUB_OUTPUT
```
Comment (avivkeller) [inline .github/workflows/sonar.yaml]:
```suggestion
          echo "SONAR_PR_NUM=$(cat pr-artifact/pr_num.txt | tr -d '\n' | tr -d ' ')" >> $GITHUB_OUTPUT
          echo "SONAR_BASE=$(cat pr-artifact/base.txt | tr -d '\n' | tr -d ' ')" >> $GITHUB_OUTPUT
          echo "SONAR_HEAD=$(cat pr-artifact/head.txt | tr -d '\n' | tr -d ' ')" >> $GITHUB_OUTPUT
```

**Review by avivkeller — APPROVED** [review]

**Review by cliffcolvin — APPROVED** [review]
LGTM, TYVM!!!

**Pull request merged**